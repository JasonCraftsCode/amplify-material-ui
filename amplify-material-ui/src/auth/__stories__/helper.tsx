import * as React from 'react';
import { AuthContext } from '../auth-context';

export const withContext = ({
    authState,
    onStateChange = () => {},
}) => component => (
    <AuthContext.Provider value={{ authState, onStateChange }}>
        {component}
    </AuthContext.Provider>
);
