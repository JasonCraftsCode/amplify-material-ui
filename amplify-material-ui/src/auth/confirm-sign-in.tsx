import * as React from 'react';
import {
    Button,
    Grid,
    Link,
    TextField,
    makeStyles,
    createStyles,
    Theme,
} from '@material-ui/core';
import Auth from '@aws-amplify/auth';
import { I18n } from '@aws-amplify/core';

import AuthProps from './auth-props';
import { FormSection, SectionHeader, SectionBody, SectionFooter } from '../ui';
import { useForm } from '../hooks';

export interface ConfirmSignInProps extends AuthProps {
    validationData?: { [key: string]: string };
    authData: any;
}

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        form: {
            width: '100%', // Fix IE 11 issue.
            marginTop: theme.spacing(1),
        },
        submit: {
            margin: theme.spacing(3, 0, 2),
        },
    }),
);

const useMfaType = (user: any) => {
    const [mfaType, setMfaType] = React.useState('SMS');

    React.useEffect(() => {
        const newMfaType =
            user && user.challengeName === 'SOFTWARE_TOKEN_MFA'
                ? 'TOTP'
                : 'SMS';
        if (mfaType !== newMfaType) {
            setMfaType(newMfaType);
        }
    }, [mfaType, user]);

    return mfaType;
};

export const ConfirmSignIn: React.FC<ConfirmSignInProps> = props => {
    const { authState, onStateChange, authData: user } = props;

    const classes = useStyles();

    const mfaType = useMfaType(user);

    const confirm = async ({ code }: { code: string }) => {
        if (!Auth || typeof Auth.confirmSignIn !== 'function') {
            throw new Error(
                'No Auth module found, please ensure @aws-amplify/auth is imported',
            );
        }
        await Auth.confirmSignIn(
            user,
            code,
            mfaType === 'TOTP' ? 'SOFTWARE_TOKEN_MFA' : null,
        );
    };

    const { inputs, handleInputChange, handleSubmit } = useForm(confirm, {
        code: '',
    });

    if (!['confirmSignIn'].includes(authState)) {
        return null;
    }

    return (
        <FormSection>
            <SectionHeader>
                {I18n.get('Confirm ' + mfaType + ' Code')}
            </SectionHeader>
            <form onSubmit={handleSubmit} className={classes.form} noValidate>
                <SectionBody>
                    <TextField
                        variant="outlined"
                        margin="normal"
                        required
                        fullWidth
                        id="code"
                        label={`${I18n.get('Code')} *`}
                        name="code"
                        autoComplete="code"
                        autoFocus
                        onChange={handleInputChange}
                        value={inputs.code}
                    />
                </SectionBody>
                <SectionFooter>
                    <Button
                        type="submit"
                        fullWidth
                        variant="contained"
                        color="primary"
                        className={classes.submit}
                    >
                        {I18n.get('Confirm')}
                    </Button>
                    <Grid container>
                        <Grid item xs>
                            <Link
                                href="#"
                                onClick={() => onStateChange('signIn', null)}
                                variant="body2"
                            >
                                {I18n.get('Back to Sign In')}
                            </Link>
                        </Grid>
                    </Grid>
                </SectionFooter>
            </form>
        </FormSection>
    );
};
