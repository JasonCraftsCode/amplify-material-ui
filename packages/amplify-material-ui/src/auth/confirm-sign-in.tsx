import * as React from 'react';
import invariant from 'tiny-invariant';
import {
    Button,
    Grid,
    makeStyles,
    createStyles,
    Theme,
} from '@material-ui/core';
import Auth from '@aws-amplify/auth';
import { I18n, JS } from '@aws-amplify/core';
import { Formik, Field, Form } from 'formik';
import { TextField } from 'formik-material-ui';

import { useAuthContext } from './auth-context';
import { ChangeAuthStateLink } from './change-auth-state-link';
import { FormSection, SectionHeader, SectionBody, SectionFooter } from '../ui';

export const useMfaType = (user: { challengeName: string }): string => {
    const [mfaType, setMfaType] = React.useState('SMS');

    React.useEffect(() => {
        const newMfaType =
            user && user.challengeName === 'SOFTWARE_TOKEN_MFA'
                ? 'TOTP'
                : 'SMS';
        if (mfaType !== newMfaType) {
            setMfaType(newMfaType);
        }
    }, [mfaType, user]);

    return mfaType;
};

export const useConfirmSignIn = (
    mfaType: string,
): ((code: string) => Promise<void>) => {
    const { onStateChange, authData: user } = useAuthContext();

    const checkContact = async (user: any): Promise<void> => {
        invariant(
            Auth && typeof Auth.verifiedContact === 'function',
            'No Auth module found, please ensure @aws-amplify/auth is imported',
        );

        const data = await Auth.verifiedContact(user);

        if (!JS.isEmpty(data.verified)) {
            onStateChange('signedIn', user);
        } else {
            const newUser = Object.assign(user, data);
            onStateChange('verifyContact', newUser);
        }
    };

    return async (code: string): Promise<void> => {
        invariant(
            Auth && typeof Auth.confirmSignIn === 'function',
            'No Auth module found, please ensure @aws-amplify/auth is imported',
        );

        try {
            await Auth.confirmSignIn(
                user,
                code,
                mfaType === 'TOTP' ? 'SOFTWARE_TOKEN_MFA' : null,
            );
            checkContact(user);
        } catch (error) {
            console.log(error);
        }
    };
};

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        form: {
            width: '100%', // Fix IE 11 issue.
            marginTop: theme.spacing(1),
        },
        submit: {
            margin: theme.spacing(3, 0, 2),
        },
    }),
);

export const ConfirmSignIn: React.FC = () => {
    const { authData: user } = useAuthContext();

    const classes = useStyles();

    const mfaType = useMfaType(user);

    const confirmSignIn = useConfirmSignIn(mfaType);

    return (
        <Formik<{ code: string }>
            initialValues={{
                code: '',
            }}
            onSubmit={async ({ code }, { setSubmitting }): Promise<void> => {
                await confirmSignIn(code);
                setSubmitting(false);
            }}
        >
            {({ handleSubmit, isValid }): React.ReactNode => (
                <FormSection>
                    <SectionHeader>
                        {I18n.get('Confirm ' + mfaType + ' Code')}
                    </SectionHeader>
                    <Form className={classes.form} onSubmit={handleSubmit}>
                        <SectionBody>
                            <Field
                                variant="outlined"
                                margin="normal"
                                required
                                fullWidth
                                id="code"
                                label={`${I18n.get('Code')} *`}
                                name="code"
                                autoComplete="code"
                                autoFocus
                                component={TextField}
                            />
                        </SectionBody>
                        <SectionFooter>
                            <Button
                                type="submit"
                                disabled={!isValid}
                                fullWidth
                                variant="contained"
                                color="primary"
                                className={classes.submit}
                            >
                                {I18n.get('Confirm')}
                            </Button>
                            <Grid container>
                                <Grid item xs>
                                    <ChangeAuthStateLink
                                        label={I18n.get('Back to Sign In')}
                                        newState="signIn"
                                    />
                                </Grid>
                            </Grid>
                        </SectionFooter>
                    </Form>
                </FormSection>
            )}
        </Formik>
    );
};
