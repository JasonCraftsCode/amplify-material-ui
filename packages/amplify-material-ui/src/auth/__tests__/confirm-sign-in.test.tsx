import * as React from 'react';
import { renderHook } from '@testing-library/react-hooks';
import { render } from '@testing-library/react';

import { withContext } from './helper';
import { ConfirmSignIn, useMfaType } from '../confirm-sign-in';

describe('confirm-sign-in', () => {
    it('should be rendered correctly in the confirmSignIn authState', () => {
        const { asFragment } = render(withContext(<ConfirmSignIn />)());
        expect(asFragment()).toMatchSnapshot();
    });
});

describe('use-mfa-type', () => {
    it('should return SMS as default', () => {
        const user = {
            challengeName: 'XXX',
        };

        const { result } = renderHook(() => useMfaType(user));

        expect(result.current).toBe('SMS');
    });

    it('should return TOTP is challengeName = SOFTWARE_TOKEN_MFA', () => {
        const user = {
            challengeName: 'SOFTWARE_TOKEN_MFA',
        };

        const { result } = renderHook(() => useMfaType(user));

        expect(result.current).toBe('TOTP');
    });
});
