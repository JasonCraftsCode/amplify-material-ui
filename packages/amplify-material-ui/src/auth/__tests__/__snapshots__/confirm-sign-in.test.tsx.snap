// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`confirm-sign-in should be rendered correctly in the confirmSignIn authState 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthXs"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation1 makeStyles-paper-3 MuiPaper-rounded"
    >
      <div
        class="MuiBox-root MuiBox-root-41 makeStyles-box-38"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-39 MuiAvatar-colorDefault"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <g
              fill="none"
            >
              <path
                d="M0 0h24v24H0V0z"
              />
              <path
                d="M0 0h24v24H0V0z"
                opacity=".87"
              />
            </g>
            <path
              d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zM9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9V6zm9 14H6V10h12v10zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2z"
            />
          </svg>
        </div>
        <h1
          class="MuiTypography-root MuiTypography-h6"
        >
          Confirm TOTP Code
        </h1>
      </div>
      <form
        action="#"
        class="makeStyles-form-1"
      >
        <div
          class="MuiBox-root MuiBox-root-88 makeStyles-box-87"
        >
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-focused Mui-focused Mui-required Mui-required"
              data-shrink="true"
              for="code"
              id="code-label"
            >
              Code *
              <span
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth Mui-focused Mui-focused MuiInputBase-formControl"
            >
              <input
                aria-invalid="false"
                autocomplete="code"
                class="MuiInputBase-input MuiOutlinedInput-input"
                id="code"
                name="code"
                required=""
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-151 MuiOutlinedInput-notchedOutline"
                style="padding-left: 8px;"
              >
                <legend
                  class="PrivateNotchedOutline-legend-152"
                  style="width: 0px;"
                >
                  <span>
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-154 makeStyles-box-153"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-submit-2 MuiButton-containedPrimary MuiButton-fullWidth"
            tabindex="0"
            type="submit"
          >
            <span
              class="MuiButton-label"
            >
              Confirm
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <div
            class="MuiGrid-root MuiGrid-container"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
            >
              <a
                class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-body2 MuiTypography-colorPrimary"
                href="#"
              >
                Back to Sign In
              </a>
            </div>
          </div>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
